IdentityCore/src/network/request_serializer/MSIDUrlRequestSerializer.m

```diff
@@ -26,30 +26,37 @@
 
 @implementation MSIDUrlRequestSerializer
 
-- (NSURLRequest *)serializeWithRequest:(NSURLRequest *)request parameters:(NSDictionary *)parameters
+- (NSURLRequest *)serializeWithRequest:(NSURLRequest *)request parameters:(NSDictionary *)parameters headers:(NSDictionary *)headers
 {
     NSParameterAssert(request);
-    
-    if (!parameters) return request;
-    
     NSMutableURLRequest *mutableRequest = [request mutableCopy];
+    NSMutableDictionary *requestHeaders = [NSMutableDictionary new];
     
-    if ([self shouldEncodeParametersInURL:request])
+    if ([headers count])
     {
-        NSAssert(mutableRequest.URL, NULL);
-        
-        NSURLComponents *urlComponents = [NSURLComponents componentsWithURL:mutableRequest.URL resolvingAgainstBaseURL:NO];
-        NSMutableDictionary *urlParameters = [[mutableRequest.URL msidQueryParameters] mutableCopy] ?: [NSMutableDictionary new];
-        [urlParameters addEntriesFromDictionary:parameters];
-        urlComponents.percentEncodedQuery = [urlParameters msidURLEncode];
-        mutableRequest.URL = urlComponents.URL;
+        [requestHeaders addEntriesFromDictionary:headers];
     }
-    else
+    
+    if ([parameters count])
     {
-        mutableRequest.HTTPBody = [[parameters msidWWWFormURLEncode] dataUsingEncoding:NSUTF8StringEncoding];
-        [mutableRequest setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
+       if ([self shouldEncodeParametersInURL:request])
+       {
+           NSAssert(mutableRequest.URL, NULL);
+           
+           NSURLComponents *urlComponents = [NSURLComponents componentsWithURL:mutableRequest.URL resolvingAgainstBaseURL:NO];
+           NSMutableDictionary *urlParameters = [[mutableRequest.URL msidQueryParameters] mutableCopy] ?: [NSMutableDictionary new];
+           [urlParameters addEntriesFromDictionary:parameters];
+           urlComponents.percentEncodedQuery = [urlParameters msidURLEncode];
+           mutableRequest.URL = urlComponents.URL;
+       }
+       else
+       {
+           mutableRequest.HTTPBody = [[parameters msidWWWFormURLEncode] dataUsingEncoding:NSUTF8StringEncoding];
+           [requestHeaders setObject:@"application/x-www-form-urlencoded" forKey:@"Content-Type"];
+       }
     }
     
+    mutableRequest.allHTTPHeaderFields = requestHeaders;
```

Comment:
Can we make sure that we don't override headers that are already set in `request`?

Reply:
Sure. I will test this and address it in a separate PR.